webcrypto/algorithm.go

```diff
@@ -160,46 +134,48 @@ func Hasher(algorithm HashAlgorithmIdentifier) (func() hash.Hash, error) {
 	return nil, NewError(0, ImplementationError, fmt.Sprintf("unsupported hash algorithm: %s", algorithm))
 }
 
+// NormalizedAlgorithm represents a normalized algorithm.
+type NormalizedAlgorithm interface {
+	NormalizedName() AlgorithmIdentifier
+}
+
 // normalize algorithm
 // normalizeAlgorithm(algorithm: string | Algorithm, op: string):
 
 // NormalizeAlgorithm normalizes the given algorithm following the algorithm described in the WebCrypto [specification].
 //
 // [specification]: https://www.w3.org/TR/WebCryptoAPI/#algorithm-normalization-normalize-an-algorithm
-func NormalizeAlgorithm(algorithm interface{}, op OperationIdentifier) (interface{}, error) {
-	// var initialAlg AlgorithmIdentifier
-	var initialAlg map[string]interface{}
-
-	switch alg := algorithm.(type) {
-	case string:
-		return NormalizeAlgorithm(map[string]interface{}{"name": alg}, op)
-	case map[string]interface{}:
-		// FIXME: this should call the NewAlgorithmFrom method instead
-		if _, ok := alg["name"]; !ok {
-			return Algorithm{}, NewError(0, SyntaxError, "algorithm name is not found")
-		}
-
-		if _, ok := alg["name"].(string); !ok {
-			return Algorithm{}, NewError(0, SyntaxError, "algorithm name is not a string")
-		}
-
-		initialAlg = alg
-	default:
-		return Algorithm{}, NewError(0, ImplementationError, "unsupported algorithm type")
-	}
-
+func NormalizeAlgorithm(rt *goja.Runtime, algorithm goja.Value, op OperationIdentifier) (NormalizedAlgorithm, error) {
 	// 1.
 	registeredAlgorithms, ok := supportedAlgorithms[op]
 	if !ok {
 		return Algorithm{}, NewError(0, ImplementationError, fmt.Sprintf("unsupported operation: %s", op))
 	}
 
-	// 2. 3. 4.
-	algName, ok := initialAlg["name"].(string)
-	if !ok {
-		return Algorithm{}, NewError(0, SyntaxError, "algorithm name is not a string")
+	// 2.
+	var initialAlg Algorithm
+	switch algorithm.ExportType().Kind() {
```

Comment:
I am not really a fan of using `ExportType` as IME it is really ... fragile.

Usually I would recommend actually checking the prototype of the value, but this is fine for now.